osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs

```diff
@@ -331,6 +332,7 @@ private void handleTimingPoints(Beatmap beatmap, string line)
 
             if (speedMultiplier != difficultyPoint.SpeedMultiplier)
             {
+                beatmap.ControlPointInfo.DifficultyPoints.RemoveAll(x => x.Time == time);

```

Comment:
How about doing this for every other control point as well? Maybe even as a post-processing step - remove all duplicates.

Other option is to add them all into a HashSet and then AddRange from that HashSet to the ControlPointInfo lists at the very end.